grpc-sys/grpc_wrap.cc

```diff
@@ -845,4 +849,122 @@ grpcwrap_ssl_server_credentials_create(
   return creds;
 }
 
+using grpc::CompletionQueue;
+
+/** Wrapper of grpc::CompletionQueue.
+ */
+typedef struct grpcwrap_completion_queue {
+  CompletionQueue* cq;
+} grpcwrap_completion_queue;
+
+/** Create a ComletionQueue, a shadow of grpc_completion_queue.
+ */
+GPR_EXPORT grpcwrap_completion_queue *GPR_CALLTYPE
+grpcwrap_completion_queue_shadow(grpc_completion_queue *cq) {
+  void *ptr = gpr_malloc(sizeof(CompletionQueue));
+  CompletionQueue *shadow_cq = new(ptr) CompletionQueue(cq);
+
+  grpcwrap_completion_queue *cq_wrapper = (grpcwrap_completion_queue *)
+    gpr_malloc(sizeof(grpcwrap_completion_queue));
+  cq_wrapper->cq = shadow_cq;
+
+  return cq_wrapper;
+}
+
+using grpc::Alarm;
+
+typedef Alarm grpcwrap_alarm;
+
+/** Create a completion queue alarm instance */
+GPR_EXPORT grpcwrap_alarm *GPR_CALLTYPE
+grpcwrap_alarm_create(void) {
+  void *ptr = gpr_malloc(sizeof(Alarm));
+  auto alarm = new(ptr) Alarm();
+  return alarm;
+}
+
+/** Set a completion queue alarm instance associated to a cq.
+ *
+ * Once the alarm notifies (see a grpcwrap_alarm_notify) or it's destroy (see a
```

Comment:
You mention `grpcwrap_alarm_notify` but I do not see anything called this. Can you help me find it?

```
hoverbear@nomad:git/grpc-rs ‹upgrade-1.10-1*›$ rg grpcwrap_alarm_notify
grpc-sys/grpc_wrap.cc
888: * Once the alarm notifies (see a grpcwrap_alarm_notify) or it's destroy (see a
```

Reply:
Hm, it should be `grpcwrap_alarm_cancel`. :man_facepalming: